frontend/packages/dev-console/src/components/pipelineruns/logs/Logs.tsx

```diff
@@ -0,0 +1,123 @@
+import { WSFactory } from '@console/internal/module/ws-factory';
+import { LOG_SOURCE_TERMINATED } from '@console/internal/components/utils';
+import { coFetchText } from '@console/internal/co-fetch';
+import './Logs.scss';
+
+type LogsProps = {
+  resource: PodKind;
+  resourceStatus: string;
+  container: ContainerSpec;
+  render: boolean;
+  autoScroll: boolean;
+  onComplete: (containerName: string) => void;
+};
+
+const Logs: React.FC<LogsProps> = ({
+  resource,
+  resourceStatus,
+  container,
+  onComplete,
+  render,
+  autoScroll,
+}) => {
+  const { name } = container;
+  const { kind, metadata = {} } = resource;
+  const { name: resName, namespace: resNamespace } = metadata;
+  const contentRef = React.useRef<HTMLDivElement>(null);
+  const [fetching, setFetching] = React.useState<boolean>(false);
+  const [error, setError] = React.useState<boolean>(false);
+  const appendMessage = React.useCallback(
+    (blockContent) => {
+      const contentLines = blockContent.split('\n').filter((line) => !!line);
+      if (contentRef.current && contentLines.length >= 0) {
+        const elements = contentLines.map((content) => {
+          const customElement = document.createElement('div');
+          customElement.textContent = content;
+          return customElement;
+        });
+        elements.forEach((element) => {
+          contentRef.current.append(element);
+        });
+        const lastElement = elements[elements.length - 1];
+        if (render && lastElement && autoScroll) {
+          lastElement.scrollIntoView({ behavior: 'smooth', block: 'end' });
+        }
+      }
+    },
+    [autoScroll, render],
+  );
+
+  React.useEffect(() => {
+    let ws: WSFactory;
+    if (!fetching) {
+      const urlOpts = {
+        ns: resNamespace,
+        name: resName,
+        path: 'log',
+        queryParams: {
+          container: name,
+          follow: 'true',
+        },
+      };
+      const watchURL = resourceURL(modelFor(kind), urlOpts);
+      if (resourceStatus === LOG_SOURCE_TERMINATED) {
+        coFetchText(watchURL)
+          .then((res) => {
+            appendMessage(res);
+            onComplete(name);
+          })
+          .catch(() => {
+            setError(true);
+            onComplete(name);
+          });
+      } else {
+        const wsOpts = {
+          host: 'auto',
+          path: watchURL,
+          subprotocols: ['base64.binary.k8s.io'],
+        };
+        ws = new WSFactory(watchURL, wsOpts);
+        ws.onmessage((msg) => {
+          const message = Base64.decode(msg);
+          appendMessage(message);
+        })
+          .onclose(() => {
+            onComplete(name);
+          })
+          .onerror(() => {
+            setError(true);
+            onComplete(name);
+          });
+      }
+      setFetching(true);
+    }
+    return () => {
+      ws && ws.destroy();
+    };
+  }, [appendMessage, fetching, kind, name, onComplete, resName, resNamespace, resourceStatus]);
```

Comment:
Store `appendMessage` and `onComplete` in `useRef` because you don't want those to cause this effect to re-run.
You shouldn't be checking `fetching` in the effect because as soon as you do fetch, you end up calling `setFetching(true)` which will clean up the old effect and call `ws.destroy()`.

Not sure how `resourceStatus` works. But what triggers a change in `resourceStatus`? Is it possible it changes when you still need the websocket streaming? Because similar to `fetching` it will destroy it.

Reply:
resource status is changed once the logs are in the waiting state or running or terminated state. I used a ref for the resource status as well.